lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart

```diff
@@ -0,0 +1,181 @@
+import 'dart:async';
+
+import 'package:cloud_firestore/cloud_firestore.dart';
+import 'package:farmsmart_flutter/model/entities/EntityCollectionInterface.dart';
+import 'package:farmsmart_flutter/model/entities/ProfileEntity.dart';
+import 'package:farmsmart_flutter/model/repositories/profile/ProfileRepositoryInterface.dart';
+import 'package:firebase_auth/firebase_auth.dart';
+
+import 'DocumentToProfileEntityTransformer.dart';
+import 'ProfileEntityToDocumentTransformer.dart';
+
+class _Fields {
+  static const collectionName = "fs_users";
+  static const ownerID = "owner";
+  static const profiles = "profiles";
+  static const currentProfile = "currentProfile";
+  static const seporator = "/";
+}
+
+class FirebaseProfileRepository implements ProfileRepositoryInterface {
+  final Firestore _firestore;
+  final FirebaseAuth _auth;
+  final _transformToFirebase = ProfileEntityToDocumentTransformer();
+  final _transformFromFirebase = DocumentToProfileEntityTransformer();
+  final _currentProfileController = StreamController<ProfileEntity>.broadcast();
+  final _profilesListController =
+      StreamController<List<ProfileEntity>>.broadcast();
+  Future<FirebaseUser> _user;
+  FirebaseProfileRepository(Firestore firestore, FirebaseAuth auth)
+      : this._firestore = firestore,
+        this._auth = auth {
+    init();
+  }
+
+  init() {
+    _user = _auth.currentUser().then((user) {
+      return user;
+    });
+    _auth.onAuthStateChanged.listen((user) {
+      _user = Future.value(user);
+    });
+  }
+
+  @override
+  Future<ProfileEntity> add(ProfileEntity profile) {
+    return _user.then((user) {
+      final Map<String, dynamic> data =
+          _transformToFirebase.transform(from: profile);
+      return _profilesCollection(user).add(data).then((result) {
+        return result.get().then((document) {
+          final newProfile = _transformFromFirebase.transform(from: document);
+          return newProfile;
```

Comment:
Why creating this final variable if you just can:

return _transformFromFirebase.transform(from: document);

Reply:
👍

Reply:
sorted 👍